flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -184,26 +188,32 @@ class JobManager(
       }
       else {
         try {
-          val instanceID = instanceManager.registerTaskManager(
-            taskManager,
-            connectionInfo,
-            hardwareInformation,
-            numberOfSlots,
-            leaderSessionID)
+        
+          if(jobManagerVersionID == taskManagerVersionID) {
+            val instanceID = instanceManager.registerTaskManager(
+              taskManager,
+              connectionInfo,
+              hardwareInformation,
+              numberOfSlots,
+              leaderSessionID)
 
-          // IMPORTANT: Send the response to the "sender", which is not the
-          //            TaskManager actor, but the ask future!
-          sender() ! decorateMessage(
-            AcknowledgeRegistration(
-              registrationSessionID,
-              leaderSessionID.get,
-              self,
-              instanceID,
-              libraryCacheManager.getBlobServerPort)
-          )
+            // IMPORTANT: Send the response to the "sender", which is not the
+            //            TaskManager actor, but the ask future!
+            sender() ! decorateMessage(
+              AcknowledgeRegistration(
+                registrationSessionID,
+                leaderSessionID.get,
+                self,
+                instanceID,
+                libraryCacheManager.getBlobServerPort,
+                jobManagerVersionID)
+            )
 
-          // to be notified when the taskManager is no longer reachable
-          context.watch(taskManager)
+            // to be notified when the taskManager is no longer reachable
+            context.watch(taskManager)
+          } else{
+            throw new Exception("Version mismatch error between Job Manager and Task Manager")
```

Comment:
This will kill the `JobManager`. This is not a good solution. Best you revert the verification between the `TaskManager` and the `JobManager`.